static/js/message_list_view.js

```diff
@@ -1445,4 +1445,87 @@ export class MessageListView {
+                visible_top + header_props.height + margin_between_recipient_rows;
+            /* If two headers next to each other are completely visible
+               (message header at top has no visible content), we don't mind showing
+               date on any of them. Which header is chosen will depend on which
+               comes first when iterating on the headers. */
+            return (
+                header_about_to_be_partially_hidden <= header_props.top &&
+                header_props.top <= header_is_sticky_or_about_to_be_sticky
+            );
+        }
+
+        const $table = rows.get_table(this.table_name);
+        const $headers = $table.find(".message_header");
+        const iterable_headers = $headers.toArray();
+        /* To get to the sticky header ASAP, we take help from how far has user scrolled. */
+        if ($message_viewport.scrollTop() > $message_viewport.height() / 2) {
+            iterable_headers.reverse();
+        }
+        let sticky_header;
+        for (const header of iterable_headers) {
+            if (is_sticky(header)) {
+                sticky_header = header;
+                break;
+            }
+        }
+
+        /* Set correct date for the sticky_header. */
+        let $message_row;
+        if (!sticky_header) {
+            /* If user at the top of the scroll container, there is no sticky
+               header, so we just set date of the header to the first visible message. */
+            sticky_header = $headers[0];
+            $message_row = $($headers[0].nextElementSibling);
+        } else {
+            $(sticky_header).addClass("sticky_header");
+            const sticky_header_props = sticky_header.getBoundingClientRect();
+            /* Get `message_row` under the sticky header. */
+            const elements_below_sticky_header = document.elementsFromPoint(
+                sticky_header_props.left,
+                sticky_header_props.top,
+            );
+            let message_row = elements_below_sticky_header.filter((element) =>
+                element.classList.contains("message_row"),
+            );
+            if (!message_row.length) {
+                /* If there is no message row under the header, it means it is not sticky yet,
+                   so we just get the message next to the header. */
+                message_row = [sticky_header.nextElementSibling];
+            }
+            $message_row = $(message_row[0]);
+        }
+        const msg_id = rows.id($message_row);
+        if (msg_id === undefined) {
+            return;
+        }
+        const message = message_store.get(msg_id);
+        if (!message) {
+            return;
+        }
+        const time = new Date(message.timestamp * 1000);
+        const today = new Date();
+        const rendered_date = timerender.render_date(time, undefined, today)[0].outerHTML;
+        $(sticky_header).find(".recipient_row_date").html(rendered_date);
```

Comment:
I know this code was copied, but avoid the `.outerHTML` and `.html()` round trip—just `.append()` the jQuery object directly.

Reply:
Okay, directly used `html` since we needed to replace the content and I didn't want to use `.empty().append()`.